main_test.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/sirupsen/logrus"
+)
+
+// TestParseEnvironmentVariables ensures that we properly parse and validate all
+// required environment variables.
+func TestParseEnvironmentVariables(t *testing.T) {
+	logger := logrus.New()
+
+	// Missing PORTAL_DOMAIN
+	err := os.Setenv(envPortal, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if err == nil || !strings.Contains(err.Error(), "missing env var "+envPortal) {
+		t.Fatal("Failed to error out on invalid", envPortal)
+	}
+	err = os.Setenv(envPortal, "siasky.net")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Missing SERVER_DOMAIN
+	err = os.Setenv(envServerDomain, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if email.ServerLockID != database.PortalName {
+		t.Fatalf("Expected ServerLockID to be %s, got %s", database.PortalName, email.ServerLockID)
+	}
+	err = os.Setenv(envServerDomain, "test.siasky.net")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Invalid ACCOUNTS_JWT_TTL
+	err = os.Setenv(envJWTTTL, "sadf")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if err == nil || !strings.Contains(err.Error(), "failed to parse env var "+envJWTTTL) {
+		t.Fatal("Failed to error out on invalid", envJWTTTL)
+	}
+	// Zero ACCOUNTS_JWT_TTL
+	err = os.Setenv(envJWTTTL, "0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if err == nil || !strings.Contains(err.Error(), "env var is set to zero, which is an invalid value (must be positive or unset)") {
+		t.Fatal("Failed to error out on zero", envJWTTTL)
+	}
+	err = os.Setenv(envJWTTTL, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Missing ACCOUNTS_EMAIL_URI
+	err = os.Setenv(envEmailURI, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if err == nil || !strings.Contains(err.Error(), envEmailURI+" is empty") {
+		t.Fatal("Failed to error out on empty", envEmailURI)
+	}
+	// Invalid ACCOUNTS_EMAIL_URI
+	err = os.Setenv(envEmailURI, "this is not empty but it's also not valid")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parseEnvironmentVariables(logger)
+	if err == nil || !strings.Contains(err.Error(), "invalid email URI") {
+		t.Log(err)
+		t.Fatal("Failed to error out on invalid", envEmailURI)
+	}
+	err = os.Setenv(envEmailURI, "smtps://disabled@example.net:not-a-password@smtp.gmail.com:465/?skip_ssl_verify=false")
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestLoadDBCredentials ensures that we validate that all required environment
+// variables are present.
+func TestLoadDBCredentials(t *testing.T) {
+	// Missing user
+	err := os.Unsetenv(envDBUser)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = loadDBCredentials()
+	if err == nil || !strings.Contains(err.Error(), "missing env var "+envDBUser) {
+		t.Log(err)
+		t.Fatal("Failed to error out on missing", envDBUser)
+	}
+	err = os.Setenv(envDBUser, "user")
```

Comment:
Where are these values e.g. `"user"` coming from? The test should be setting the env vars back to what they were before unsetting them, in a `defer` so that it happens if the test errors out, as well. Also, we should make sure that this test doesn't run in parallel with other tests (I think it doesn't, but have forgotten a lot about Go already).

Reply:
This is a good point. We should make sure no other test will be influenced by this. The test shouldn't run in parallel with other tests but I'm not sure if every test is an isolated process with its own set of env vars.

Reply:
Restoring the values is a good point. It doesn't have any direct effect but it's the right thing to do.

I am not worried about this test affecting any other test because no other test should be relying on the same env vars - they are supposed to be read only in `main`. That said, I don't think I have any way of preventing tests from other packages from being affected (go runs tests within the same package sequentially but runs the tests for different packages in parallel). If that happens I'll just remove this test - it's not terribly useful anyway.

The hardcoded test values are just that - test values. But I can just as well use the original values. That might reduce the confusion.